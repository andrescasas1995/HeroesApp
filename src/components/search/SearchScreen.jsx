import React, { useMemo } from 'react';
import queryString from 'query-string'
import { HeroCard } from '../heroes/HeroCard';
import { useForm } from '../../hooks/useForm';
import { useLocation } from 'react-router-dom';
import { getHeroByName } from '../../selectors/getHeroByName';

export const SearchScreen = ({history}) => {
    const location = useLocation();
    const { q = '' } = queryString.parse(location.search);
    const [values, handleInputChange] = useForm({
        search: q
    });
    const { search } = values;
    const heroesFiltered = useMemo(() => getHeroByName(q), [q]);

    const handleSubmit = (e) => {
        e.preventDefault();
        history.push(`?q=${search}`);
    }

    return (
        <div>
            <h1>Search Screen</h1>
            <hr />

            <div className="row">
                <div className="col-5">
                    <h4>Search Form</h4>
                    <hr />

                    <form onSubmit={handleSubmit}>
                        <input
                            type="text"
                            name="search"
                            placeholder="Find your hero"
                            className="form-control"
                            autoComplete="off"
                            value={search}
                            onChange={handleInputChange}
                        />

                        <button
                            className="btn mt-1 btn-block btn-outline-primary"
                        >
                            Search
                        </button>
                    </form>
                </div>
                    <div className="col-7">
                        <h4>Results</h4>
                        <hr />

                        {
                            q === '' && (
                                <div className="alert alert-info animate__animated animate__fadeIn">
                                    Search a hero
                                </div>
                            )
                        }

                        {
                            q !== '' && heroesFiltered.length === 0 && (
                                <div className="alert alert-warning animate__animated animate__fadeIn">
                                    There is not a hero with <b>" {q} "</b>
                                </div>
                            )
                        }

                        {
                            heroesFiltered.map(hero => (
                                <div key={hero.id} className="animate__animated animate__fadeIn">
                                    <HeroCard {...hero} />
                                </div>
                            ))
                        }
                    </div>
            </div>
        </div>
    )
}
