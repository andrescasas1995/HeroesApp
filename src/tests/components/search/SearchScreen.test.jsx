import React from 'react';
import { mount } from 'enzyme';
import { MemoryRouter, Route } from 'react-router-dom';
import { SearchScreen } from '../../../components/search/SearchScreen';

describe('Pruebas en el componente SearchScreen', () => {
    test('Debe de mostrarse correctamente con valores por defecto', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <Route path="/search" component={SearchScreen} />
            </MemoryRouter>
        );

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert-info').text().trim()).toBe('Search a hero');
    });
    
    test('Debe de mostrar a Batman y el el query con el string', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <Route path="/search" component={SearchScreen} />
            </MemoryRouter>
        );

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('input').prop('value')).toBe('batman');
    });
    
    test('Debe de mostrar un error si no se encuentra el hero', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batmant']}>
                <Route path="/search" component={SearchScreen} />
            </MemoryRouter>
        );

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert-warning').text().trim()).toBe('There is not a hero with " batmant "');
    });
    
    test('Debe de llamar el push del history', () => {
        const history = {
            push: jest.fn()
        };

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batmant']}>
                <Route
                path="/search"
                component={() => <SearchScreen history={history} />}
            />
            </MemoryRouter>
        );

        wrapper.find('input').simulate('change', {
            target: {
                name: 'search',
                value: 'thor'
            }
        });

        wrapper.find('form').prop('onSubmit')({
            preventDefault(){}
        });

        expect(history.push).toHaveBeenCalledWith(`?q=thor`);
    });
});
