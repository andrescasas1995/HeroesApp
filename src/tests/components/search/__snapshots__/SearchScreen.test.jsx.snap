// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente SearchScreen Debe de mostrar a Batman y el el query con el string 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="search"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="batman"
        />
        <button
          className="btn mt-1 btn-block btn-outline-primary"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="animate__animated animate__fadeIn"
        key="dc-batman"
      >
        <div
          className="card ms-3"
          style={
            Object {
              "maxWidth": 540,
            }
          }
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-md-4"
            >
              <img
                alt="Batman"
                className="card-img"
                src="./assets/heroes/dc-batman.jpg"
              />
            </div>
            <div
              className="col-md-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Batman
                  <img
                    alt="Batman"
                    className="card-img ml-2"
                    src="../assets/publishers/DC Comics.png"
                    style={
                      Object {
                        "maxWidth": 50,
                      }
                    }
                  />
                </h5>
                <p
                  className="card-text"
                >
                  Bruce Wayne
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  >
                    Detective Comics #27
                  </small>
                </p>
                <a
                  href="/hero/dc-batman"
                  onClick={[Function]}
                >
                  MÃ¡s...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en el componente SearchScreen Debe de mostrar un error si no se encuentra el hero 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="search"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="batmant"
        />
        <button
          className="btn mt-1 btn-block btn-outline-primary"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-warning animate__animated animate__fadeIn"
      >
        There is not a hero with 
        <b>
          " 
          batmant
           "
        </b>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en el componente SearchScreen Debe de mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="search"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <button
          className="btn mt-1 btn-block btn-outline-primary"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info animate__animated animate__fadeIn"
      >
        Search a hero
      </div>
    </div>
  </div>
</div>
`;
